### PR TITLE
ECM-199: Replace gulp with angular cli

### PR DESCRIPTION
**Pull request by larsduelfer**
Replace the gulp build in the educama frontend with angular cli and activate ahead-of-time compilation for the distribution bundle.

Since the whole directory structure changes, the commits contain a huge amount of changed files.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,11 +1,14 @@
 # Configuration file for Travis CI, see https://docs.travis-ci.com
+sudo: required
```
Comment (tobiasschaefer) [inline .travis.yml]: it's good that we're switching hard to trusty, but why do we need sudo?

File: api/checkstyle.xml [hunk 2]
```diff
@@ -0,0 +1,134 @@
+<?xml version="1.0"?>
```
Comment (tobiasschaefer) [inline api/checkstyle.xml]: can't we reference the backend/checkstyle.xml file instead of creating a copy here?

File: api/pom.xml [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	<version>0.0.1-SNAPSHOT</version>
+	<packaging>jar</packaging>
+
+	<name>educama-api-gateway</name>
+	<description>API Gateway for Educama</description>
+
+	<parent>
+		<groupId>org.springframework.boot</groupId>
+		<artifactId>spring-boot-starter-parent</artifactId>
+		<version>2.0.0.M2</version>
```
Comment (tobiasschaefer) [inline api/pom.xml]: Do we actually need a milestone release of 2.0 or is 1.5.x not enough to configure the api gateway?

File: api/src/main/resources/application.yml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    name: api
+
+# Tomcat Configuration
+server:
+  port: 8081
+
+# Default paths to services. Can be overwritten in manifests.
+educama:
+  backend:
+    url: http://localhost:8082
```
Comment (tobiasschaefer) [inline api/src/main/resources/application.yml]: why is localhost the default here? Why not the development space in the cloud?

File: backend/src/main/resources/application.properties [hunk 5]
```diff
@@ -1,5 +1,5 @@
 # Tomcat Configuration
-server.port=8081
+server.port=8082
```
Comment (tobiasschaefer) [inline backend/src/main/resources/application.properties]: why change port?

File: backend/src/main/resources/application.properties [hunk 6]
```diff
@@ -25,3 +25,6 @@ flyway.locations=classpath:db/migration/h2,classpath:db/migration/common
 #spring.jpa.properties.javax.persistence.schema-generation.scripts.action=create
 #spring.jpa.properties.javax.persistence.schema-generation.scripts.create-target=database-schema.sql
 #spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.MySQL5Dialect
+
+# Configuration of Spring Boot enpoints. Security is switched of for now.
```
Comment (tobiasschaefer) [inline backend/src/main/resources/application.properties]: typo: enpoints

File: api/src/main/resources/application.yml [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      path: /educama/v1/flights/**
+      url: ${educama.flightservice.url}/connections
+      stripPrefix: true
+
+ribbon:
+  eureka:
+    enabled: false
+
+# Configuration of Spring Boot enpoints. Security is switched of for now.
+endpoints:
```
Comment (tobiasschaefer) [inline api/src/main/resources/application.yml]: why not off completly for now: management.security.enabled = false?

File: build.sh [hunk 8]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+cd backend
```
Comment (tobiasschaefer) [inline build.sh]: The following ist better because it will switch back the working directory automatically.
( cd backend && ./build.sh)

File: build.sh [hunk 9]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
```
Comment (tobiasschaefer) [inline build.sh]: please add "set -e" at the beginning of the script so that it stops on the first error

File: frontend/Staticfile [hunk 10]
```diff
@@ -1,2 +1 @@
-# enable pushstate routing, see http://docs.cloudfoundry.org/buildpacks/staticfile/index.html
```
Comment (tobiasschaefer) [inline frontend/Staticfile]: why did you remove that comment?

File: frontend/package.json [hunk 11]
```diff
@@ -3,99 +3,67 @@
...
+    "all": "npm run clean && npm install && npm run lint && npm run prod && npm run test && npm run docs && npm run zip",
+    "clean": "rm ./node_modules -rf",
+    "serve": "ng serve",
+    "build": "ng build",
+    "prod": "ng build --prod",
+    "test": "ng test -sr --cc=true",
+    "lint": "ng lint",
+    "e2e": "ng e2e",
+    "docs": "./node_modules/.bin/compodoc -p tsconfig.json -d docs",
+    "zip": "cd dist/ && zip -r educama-frontend.zip * && cd .. && zip ./dist/educama-frontend.zip Staticfile"
```
Comment (tobiasschaefer) [inline frontend/package.json]: please also use brackets here to prevent cd.."

File: test.sh [hunk 12]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Comment (tobiasschaefer) [inline test.sh]: please add "set -e" and use brackets to prevent cd..

File: login.sh [hunk 13]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
```
Comment (tobiasschaefer) [inline login.sh]: #!/bin/bash for consistency ?

File: backend/pom.xml [hunk 14]
```diff
@@ -139,6 +145,8 @@
                         <include>**/*Test.java</include>
                         <include>**/*Documentation.java</include>
                     </includes>
+                    <!-- Travis build workaround -->
```
Comment (tobiasschaefer) [inline backend/pom.xml]: can you document the error message you got previously in the comment? It was not clealy outOfMemory, or was it?

File: .travis.yml (hunk 1 above)
Comment (larsduelfer) [inline .travis.yml]: There was a problem when installing chrome it sudo solved this. I think there migh be another option, but I could not get it running so far. So I suggest, we solve this when restructuring our build anyway. OK?

File: api/checkstyle.xml (hunk 2 above)
Comment (larsduelfer) [inline api/checkstyle.xml]: Yes, I just wanted to keep the things independent. The flight-service also has a copy of the checkstyle rules. So, independency vs. no redundancy ... What do you think?

File: api/pom.xml (hunk 3 above)
Comment (larsduelfer) [inline api/pom.xml]: I was keen to get used to the latest version. Since this is an experimental project anyways, why not chosing a milestone release? Makes migration to the latest version easier, when the final release is out. What do you think?

File: api/src/main/resources/application.yml (hunk 4 above)
Comment (larsduelfer) [inline api/src/main/resources/application.yml]: We do not use a manifest, when running locally (on localhost, not pcfdev), so to be able to start it locally, I chose this to be the default. The manifests allow to override it anyways. Maybe a good convention to define localhost as the standard and override it for each deployment. Ok?

File: api/src/main/resources/application.yml (hunk 7 above)
Comment (larsduelfer) [inline api/src/main/resources/application.yml]: Also an option. Did not really think about it. Can try it.

File: backend/pom.xml (hunk 14 above)
Comment (larsduelfer) [inline backend/pom.xml]: No, it was a message that the vm stopped working. After I googled it, I found this fix. Can document it.

File: .travis.yml (hunk 1 above)
Comment (tobiasschaefer) [inline .travis.yml]: Ok, but please add an inline comment to .travis.yml that you need sudo for the chrome installation.

File: api/checkstyle.xml (hunk 2 above)
Comment (tobiasschaefer) [inline api/checkstyle.xml]: Ok, keep it redundant.

File: api/pom.xml (hunk 3 above)
Comment (tobiasschaefer) [inline api/pom.xml]: OK, latest milestone.

File: api/src/main/resources/application.yml (hunk 4 above)
Comment (tobiasschaefer) [inline api/src/main/resources/application.yml]: ok

File: backend/src/main/resources/application.properties (hunk 5 above)
Comment (larsduelfer) [inline backend/src/main/resources/application.properties]: Hi Tobias, I have not answered on this yet. I changed the ports to user 8081 for the API Gateway. So the frontend can keep the logic for localhost to finde the local backend.

File: api/src/main/resources/application.yml (hunk 7 above)
Comment (larsduelfer) [inline api/src/main/resources/application.yml]: I changed it.

File: backend/src/main/resources/application.properties (hunk 6 above)
Comment (larsduelfer) [inline backend/src/main/resources/application.properties]: Done.

File: build.sh (hunk 9 above)
Comment (larsduelfer) [inline build.sh]: Done.

File: build.sh (hunk 8 above)
Comment (larsduelfer) [inline build.sh]: Done.

File: frontend/Staticfile (hunk 10 above)
Comment (larsduelfer) [inline frontend/Staticfile]: Is back in.

File: frontend/package.json (hunk 11 above)
Comment (larsduelfer) [inline frontend/package.json]: Done.

File: login.sh (hunk 13 above)
Comment (larsduelfer) [inline login.sh]: Done.

File: backend/pom.xml (hunk 14 above)
Comment (larsduelfer) [inline backend/pom.xml]: Done.

File: test.sh (hunk 12 above)
Comment (larsduelfer) [inline test.sh]: Done.
